### PR TITLE
#161342857 Create endpoint to report an article

### PR DESCRIPTION
**Pull request by PriscillaSam**
#### What does this PR do?
Create endpoint to report an article. **POST** `/api/v1/articles/:slug/complaints`.

#### Description of Task to be completed
- add function to create a complaint.
- add the endpoint to post a complaint.
- abstract the function to check if an article exists.
- add a function to wrap controller functions in a try catch.

#### How should this be manually tested?
- on the terminal run git clone https://github.com/andela/haven-ah-backend.git.
- run git checkout `ft-post-complaints-161342857`.
- ensure that PostgreSQL is running on your local machine.
- run `npm install` to install the npm packages.
- run `npm test` to run the unit and integration tests.

To test on postman
- run npm run start-dev and open up postman.

To test the routes above, the user needs to be authenticated.
- signup (`localhost:5000/api/v1/users/signup`) and use email and password to log in (`localhost:5000/api/v1/users/signin`). Copy token and pass to the x-access-token header to test the endpoint.

Request Body should contain
```gherkin
{
  complaintType: One of ['Rules Violation', 'Abuse', 'Plagiarism', 'Others'],
  complaintBody:  "Report plagiarism",
}
```
Sample Responses
If user is the author of the article
```gherkin
{
  status: 400,
  message: "Sorry, you can not raise a complaint against your article."
}
```

if post is successful
```gherkin
{
  status: 201,
  message:  "You have logged a complaint on this article."
}
```

#### Screenshots if applicable
screen shot 2018-11-07 at 6 11 33 pm

#### What are the relevant pivotal tracker stories?
#161342857

### CODE REVIEW

**Review by wiztemple — COMMENTED** [review]
well organized. Excellent

File: server/controllers/article.js [hunk 1]
```diff
@@ -59,18 +56,47 @@ class Article {
...
+   */
+  static async postComplaint(request, response) {
+    const { userId, article } = request;
+    const { complaintBody, complaintType } = request.body;
+
+    if (article.userid === userId) {
+      return badHttpResponse(
+        response,
+        400,
+        'What is wrong with you? Why do you want to raise a complaint against yourself',
```
Comment (ucheg6) [inline server/controllers/article.js]: This error message is too harsh lol. You can try this "Please reconsider raising a complaint against yourself".

**Comment by ucheg6** [comment]
Priscilla this is a work well done.

File: server/controllers/article.js (hunk 1 above)
Comment (PriscillaSam) [reply server/controllers/article.js]: Lol. I wasn't supposed to leave this here. Thanks. I will address it.

File: package.json [hunk 2]
```diff
@@ -30,6 +30,7 @@
     "express": "^4.16.4",
     "jsonwebtoken": "^8.3.0",
     "morgan": "^1.9.1",
+    "nock": "^10.0.2",
```
Comment (seyi-adeleke) [inline package.json]: pretty sure this package has been installed before

File: server/controllers/article.js [hunk 3]
```diff
@@ -21,34 +21,31 @@ class Article {
...
-            201,
-            'Article Created and Tags associated',
-            { newArticle, combinedTags }
-          ));
-      } else {
-        return goodHttpResponse(response, 201, 'Article Created', newArticle);
-      }
-    } catch (error) {
-      return badHttpResponse(response, 500, 'There was an internal error', error);
+
```
Comment (seyi-adeleke) [inline server/controllers/article.js:25]: Why did you remove the try/catch ?

File: server/controllers/article.js [hunk 4]
```diff
@@ -59,18 +56,47 @@ class Article {
...
+      return badHttpResponse(
+        response,
+        400,
+        'Sorry, you can not raise a complaint against your article.',
+      );
+    }
+    const complaint = await articleRepo.createArticleComplaint({
+      userId,
+      complaintBody,
+      complaintType,
```
Comment (seyi-adeleke) [inline server/controllers/article.js:91]: can we have a custom type? What if my complaint is not part of the available options?

File: server/middlewares/checkIfArticleExists.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+   * Check if an article exists
+   * @param {object} request Request Object
+   * @param {object} response Response Object
+   * @param {function} next Next function on the middleware chain
+   * @returns {object} Error is article does not exist
+   * or calls the next function
+   */
+const checkIfArticleExists = async (request, response, next) => {
+  const { slug } = request.params;
+  const article = await articleRepo.getArticleByParam('slug', slug);
```
Comment (seyi-adeleke) [inline server/middlewares/checkIfArticleExists.js]: use the getArticleBySlug function

File: server/middlewares/paramsValidator.js [hunk 6]
```diff
@@ -8,17 +8,45 @@ import { badHttpResponse } from '../utilities/httpResponse';
...
-      errors[key] = `The ${key} parameter must be a ${definition.type}`;
-    }
-  });
+      if (!isMatch) {
+        errors[key] = `The ${key} parameter provided is not a valid ${key}`;
+      }
+    });
+  }
+
+  if (bodyParams) {
```
Comment (seyi-adeleke) [inline server/middlewares/paramsValidator.js:28]: This is hard to follow and seems like an unnecessary complication. I'd recommend adding custom validation to what you're supposed to be working on and leaving the rest of the codebase as is

File: server/repository/articleRepository.js [hunk 7]
```diff
@@ -57,21 +50,43 @@ class ArticleRepository {
...
-    const singleArticle = await Articles.findOne({
-      where: {
-        slug
-      }
+  * Finds an article by parameter supplied
+  * @param {string} param Parameter to search by
+  * @param {string} value Parameter value
+  * @returns {object | null} Article object or null if article is not found
+  */
+  static async getArticleByParam(param, value) {
```
Comment (seyi-adeleke) [inline server/repository/articleRepository.js]: can this function take multiple params as opposed to one?  I prefer the  getArticleBySlug method because its simple and straight to the point.

File: server/routes/article.js [hunk 8]
```diff
@@ -2,16 +2,24 @@ import { Router } from 'express';
 import Article from '../controllers/article';
 import inputValidator from '../middlewares/validations';
 import isAuthenticated from '../middlewares/checkAuth';
+import validator from '../middlewares/paramChecker';
+import tryCatchWrapper from '../utilities/tryCatchWrapper';
```
Comment (seyi-adeleke) [inline server/routes/article.js:7]: ah. nice

**Review by seyi-adeleke — CHANGES_REQUESTED** [review]

File: server/controllers/article.js (hunk 4 above)
Comment (PriscillaSam) [reply server/controllers/article.js:91]: @seyi-adeleke yes we can

File: server/middlewares/checkIfArticleExists.js (hunk 5 above)
Comment (PriscillaSam) [reply server/middlewares/checkIfArticleExists.js]: @seyi-adeleke Noted.

File: server/repository/articleRepository.js (hunk 7 above)
Comment (PriscillaSam) [reply server/repository/articleRepository.js]: @seyi-adeleke, Victor worked on this. We thought of a situation similar to the user where we might want to get an article by different parameters. That is the reason for this generic function.

File: server/middlewares/paramsValidator.js (hunk 6 above)
Comment (PriscillaSam) [reply server/middlewares/paramsValidator.js:28]: @seyi-adeleke I initially created this function to check for url parameters but I wanted to be able to use it for the body parameters as well. I'd just leave it as it was before this pull request.

File: server/controllers/article.js (hunk 4 above)
Comment (PriscillaSam) [reply server/controllers/article.js:91]: @seyi-adeleke I added an `Others` option.

**Review by seyi-adeleke — APPROVED** [review]
